src/main.rs

```diff
@@ -436,7 +442,20 @@ extern "C" fn kernel_init(hart_id: u64, fdt_addr: u64) {
         - mem_map.regions().next().unwrap().base().bits();
 
     // Parse the user-mode ELF containing the user-mode task.
-    let umode_bytes = include_bytes!("../target/riscv64gc-unknown-none-elf/release/umode");
+    // Safe, because it comes from the Linker
+    let umode_bytes;
+    #[cfg(feature = "bazel")]
+    {
+        let umode_bin = unsafe { core::ptr::addr_of!(_umode_bin) as *const u8 };
+        let umode_bin_len = unsafe { core::ptr::addr_of!(_umode_bin_len) as usize };
+        let pre_umode_bytes = core::ptr::slice_from_raw_parts::<u8>(umode_bin, umode_bin_len);
+        umode_bytes = unsafe { pre_umode_bytes.as_ref().unwrap() };
+    }
+    #[cfg(not(feature = "bazel"))]
+    {
+        umode_bytes = include_bytes!("../target/riscv64gc-unknown-none-elf/release/umode");
```

Comment:
I think you can get the `include_bytes!()` to work in Bazel by:

(1) Use an env macro to locate the file to include at compile time:

```
umode_bytes = include_bytes!(env!("UMODE_OBJECT"));
```

(2) In the build rule that compiles this source file, add a `rust_env` parameter that gets populated with the location to the umode object:

```
    rustc_env = {
        "UMODE_OBJECT": "$(location //src:umode_to_object)",
    },
```

(There are some rules around the precise order of macro expansion, but I've successfully used this pattern in my own project.)

You could set the same environment variable in the Rust build script so you could get rid of the `#[cfg]` here too.

Reply:
One of the motivations of this was to get rid of the include_bytes. The cfg is only around as long as the two build systems co-exist.